util/network-devp2p/src/ip_utils.rs

```diff
@@ -358,6 +362,81 @@ pub fn map_external_address(local: &NodeEndpoint) -> Option<NodeEndpoint> {
 	None
 }
 
+fn search_natpmp(local: &NodeEndpoint) -> Option<NodeEndpoint> {
+	if let SocketAddr::V4(ref local_addr) = local.address {
+		let local_port = local_addr.port();
+		let local_udp_port = local.udp_port;
+
+		let search_gateway_child = ::std::thread::spawn(move || {
+			let get_public_addr = |n: &mut Natpmp| {
+				n.send_public_address_request()?;
+				::std::thread::sleep(Duration::from_millis(50));
+				n.read_response_or_retry()
+			};
+
+			let get_mapped_port = |n: &mut Natpmp, tcp_flag: bool| {
+				if tcp_flag {
+					n.send_port_mapping_request(Protocol::TCP, local_port, local_port, NAT_PMP_PORT_MAPPING_LIFETIME)?;
+				} else {
+					n.send_port_mapping_request(Protocol::UDP, local_udp_port, local_udp_port, NAT_PMP_PORT_MAPPING_LIFETIME)?;
+				}
+				::std::thread::sleep(Duration::from_millis(50));
+				n.read_response_or_retry()
+			};
+
+			match Natpmp::new() {
+				Ok(mut n) => {
+					let gw = get_public_addr(&mut n)?;
+					let tcp_r = get_mapped_port(&mut n, true)?;
+					let udp_r = get_mapped_port(&mut n, false)?;
+
+					let gw = if let Response::Gateway(gw) = gw {
+						gw
+					} else {
+						return Err(natpmp::Error::NATPMP_ERR_UNDEFINEDERROR.into())
+					};
+
+					let tcp_r = if let Response::TCP(tcp_r) = tcp_r {
+						tcp_r
+					} else {
+						return Err(natpmp::Error::NATPMP_ERR_UNDEFINEDERROR.into())
+					};
+
+					let udp_r = if let Response::UDP(udp_r) = udp_r {
+						udp_r
+					} else {
+						return Err(natpmp::Error::NATPMP_ERR_UNDEFINEDERROR.into())
+					};
+
+					Ok(NodeEndpoint { address: SocketAddr::V4(SocketAddrV4::new(*gw.public_address(), tcp_r.public_port())), udp_port: udp_r.public_port() })
+				},
+				Err(e) => Err(e)
+			}
+		});
+
+		return search_gateway_child.join()
+			.map(|node| {
+				node.map_err(|e| debug!("NAT PMP error: {:?}", e)).ok()
```

Comment:
Having a proper error here would be nice.

Reply:
done